nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java

```diff
@@ -0,0 +1,1076 @@
+                                        failUnmappedColumns, warningUnmappedColumns, escapeColumnNames, quoteTableName);
+                            } else {
+                                log.error("Statement Type {} is not valid, flowfile {} will be penalized and routed to failure", new Object[]{statementType, flowFile});
+                                flowFile = session.penalize(flowFile);
+                                session.transfer(flowFile, REL_FAILURE);
+                                return;
+                            }
+                        } catch (final ProcessException pe) {
+                            log.error("Failed to convert {} to a SQL {} statement due to {}; routing to failure",
+                                    new Object[]{flowFile, statementType, pe.toString()}, pe);
+                            flowFile = session.penalize(flowFile);
+                            session.transfer(flowFile, REL_FAILURE);
+                            return;
+                        }
+
+                        try (PreparedStatement ps = con.prepareStatement(sqlHolder.getSql())) {
+
+                            try {
+                                ps.setQueryTimeout(queryTimeout); // timeout in seconds
+                            } catch (SQLException se) {
+                                // If the driver doesn't support query timeout, then assume it is "infinite". Allow a timeout of zero only
+                                if (queryTimeout > 0) {
+                                    throw se;
+                                }
+                            }
+
+                            Record currentRecord;
+                            List<Integer> fieldIndexes = sqlHolder.getFieldIndexes();
+
+                            while ((currentRecord = recordParser.nextRecord()) != null) {
+                                Object[] values = currentRecord.getValues();
+                                if (values != null) {
+                                    if (fieldIndexes != null) {
+                                        for (int i = 0; i < fieldIndexes.size(); i++) {
+                                            ps.setObject(i + 1, values[fieldIndexes.get(i)]);
+                                        }
+                                    } else {
+                                        // If there's no index map, assume all values are included and set them in order
+                                        for (int i = 0; i < values.length; i++) {
+                                            ps.setObject(i + 1, values[i]);
+                                        }
+                                    }
+                                    ps.addBatch();
+                                }
+                            }
+
+                            log.debug("Executing query {}", new Object[]{sqlHolder});
+                            ps.executeBatch();
```

Comment:
I think we should rollback RDBMS session and add exception details to FlowFile attribute then route it to 'retry' so that it can be analyzed and retry later. As an example, I got following exception:
```
java.sql.BatchUpdateException: Duplicate entry '2' for key 'PRIMARY'
```